src/libmw/include/mw/models/tx/Kernel.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <mw/common/Macros.h>
+#include <mw/crypto/Crypto.h>
+#include <mw/crypto/Hasher.h>
+#include <mw/traits/Committed.h>
+#include <mw/traits/Hashable.h>
+#include <mw/traits/Serializable.h>
+#include <mw/traits/Printable.h>
+#include <mw/models/crypto/BlindingFactor.h>
+#include <mw/models/crypto/Signature.h>
+#include <mw/models/tx/PegOutCoin.h>
+#include <mw/models/tx/KernelType.h>
+#include <boost/optional.hpp>
+
+class Kernel :
+    public Traits::ICommitted,
+    public Traits::IHashable,
+    public Traits::ISerializable
+{
+public:
+    Kernel() = default;
+    Kernel(
+        boost::optional<uint64_t> fee,
+        boost::optional<uint64_t> pegin,
+        boost::optional<PegOutCoin> pegout,
+        boost::optional<uint64_t> lockHeight,
+        std::vector<uint8_t> extraData,
+        Commitment&& excess,
+        Signature&& signature
+    ) : m_fee(fee),
+        m_pegin(pegin),
+        m_pegout(std::move(pegout)),
+        m_lockHeight(lockHeight),
+        m_extraData(std::move(extraData)),
```

Comment:
parameter isn't xvalue

Reply:
That's actually OK, if my understanding of move semantics is correct. I didn't realize this until just a few months ago, but you can still pass an rvalue to a function that expects an lvalue, and you can then move the lvalue parameter to another object. It results in 0 object copies and 2 moves (probably less thanks to compiler optimizations). This allows us to avoid having separate constructors for different combinations of rvalue and lvalue params.

Feel free to ping me on Telegram if that doesn't make sense, or if you disagree with my conclusion.

Reply:
If rvalues are those temporary values that are the result of some constructor/function then yes I think you're correct.